mlflow/pyfunc/scoring_server/client.py

```diff
@@ -1,9 +1,32 @@
 import requests
 import time
 import json
+import numpy as np
+import pandas as pd
 
 from mlflow.pyfunc import scoring_server
-from mlflow.utils.proto_json_utils import _DateTimeEncoder
+from mlflow.utils.proto_json_utils import (
+    dataframe_from_parsed_json,
+    parse_tf_serving_input,
+)
+from mlflow.exceptions import MlflowException
+
+
+def infer_and_parse_json_output(decoded_input):
```

Comment:
NOTE: I am not sure what the client is for and what is the intended API. I tried to preserve the existing behavior as closely as possible, but maybe we should change it?

Reply:
when `mlflow.pyfunc.spark_udf()` is called with `env_manager="virtualenv"` or `env_manager="conda"`, MLflow starts a model server and uses this client to send requests to it in order to perform inference; this is the long-awaited "restore model environments in spark UDF" feature that we talked about ~ 3 years ago :). It isn't a public API, so changes are okay as long as Spark UDF continues to work.

cc @WeichenXu123

Reply:
Yes. The `scoring_server.client` module is used in `spark_udf` with env_manager="virtualenv" or env_manager="conda".